# Казанова Анастасия Станиславовна, гр. 0373.
## Практическое задание по дисциплине "Теория принятия решений".  
### Вариант 139.  


### **Задание №1.**

Запасы товара на трех складах торговой компании, составляют 148, 108, и 101. Компания имеет
пункты реализации товара в пяти населенных пунктах, причем средние транспортные затраты
на перевозку единицы товара из складов в населенные пункты, а также спрос на товар в каждом
из пунктов указаны в табл. 1.
Для стимуляции спроса торговая компания может провести рекламную акцию в каждом из
пунктов. Потенциальный эффект от рекламной акции (увеличение спроса на единицу стоимости рекламной акции) зависит от бюджета рекламной акции. Параметры зависимости различны для городов и сведены в табл. 2. Следует обратить внимание, что эффект характеризуется
убывающей отдачей (что в определенной степени моделирует насыщение рынка). Так, например, при бюджете рекламной акции 120, увеличение спроса в первом городе будет вычисляться
следущим образом: $ 0.4 \cdot 40 + 0.3 \cdot 60 + 0.2 \cdot (120 - 40 - 60) $.
Требуется:
1. Определить минимальные издержки на реализацию всего товара.
2. Выявить населенные пункты, требующие максимальный и минимальный рекламный бюджет.
3. Провести анализ чувствительности оптимальной стратегии к стоимости перевозок между
складом 2 и населенным пунктом 3.


*Табл.1*

| Склад/Город | 1  | 2  | 3  | 4  | 5  |
|-------------|----|----|----|----|----|
| 1           | 10 | 10 | 7  | 7  | 7  |
| 2           | 5  | 7  | 6  | 5  | 7  |
| 3           | 10 | 9  | 10 | 9  | 9  |
| Спрос       | 56 | 37 | 53 | 26 | 47 |

*Табл.2*

| Стоимость акции | 1   | 2   | 3   | 4   | 5   |
|-----------------|-----|-----|-----|-----|-----|
| до 40           | 0.4 | 0.6 | 0.6 | 0.6 | 0.3 |
| до 100          | 0.3 | 0.5 | 0.5 | 0.5 | 0.2 |
| более 100       | 0.2 | 0.4 | 0.4 | 0.4 | 0.1 |

In [114]:
from cvxopt.modeling import variable, op
from cvxopt import matrix, solvers

### *1.1. Определим минимальные издержки на реализацию всего товара:*

Формальная постановка:  
$ F(x) = 10x_1 + 10x_2 + 7x_3 + 7x_4 + 7x_5 + 5x_6 + 7x_7 + 6x_8 + 5x_9 + 7x_{10} + 10x_{11} + 9x_{12} + 10x_{13} + 9x_{14} + 9x_{15} \to min$

$\left\{\begin{matrix}
x_1 + x_2 + x_3 + x_4 + x_5 \le 148  \\ 
x_6 + x_7 + x_8 + x_9 + x_{10} \le 108 \\ 
x_{11} + x_{12} + x_{13} + x_{14} + x_{15} \le 101 \\ 
x_1 + x_6 + x_{11} = 56 \\ 
x_2 + x_7 + x_{12} = 37 \\
x_3 + x_8 + x_{13} = 53 \\
x_4 + x_9 + x_{14} = 26 \\
x_5 + x_{10} + x_{15} = 47 \\
x \ge 0
\end{matrix}\right. $


Имеем классическую транспортную задачу линейного программирования. 

In [115]:
x = variable(15, 'x')
F = (10 * x[0] + 10 * x[1] + 7 * x[2] + 7 * x[3] + 7 * x[4] + 
     5 * x[5]+ 7 * x[6] + 6 * x[7] + 5 * x[8] + 7 * x[9] +
     10 * x[10] + 9 * x[11] + 10 * x[12] + 9 * x[13] + 9 * x[14])

lim1 = (x[0] + x[1] + x[2] + x[3] + x[4] <= 148)
lim2 = (x[5] + x[6] + x[7] + x[8] + x[9] <= 108)
lim3 = (x[10] + x[11] + x[12] + x[13] + x[14] <= 101)
lim4 = (x[0] + x[5] + x[3] == 56)
lim5 = (x[1] +x[6] + x[11] == 37)
lim6 = (x[2] + x[7] + x[12] == 53)
lim7 = (x[3] + x[8] + x[13] == 26)
lim8 = (x[4] + x[9] + x[14] == 47)   
problem = op(F, [lim1, lim2, lim3, lim4, lim5, lim6, lim7, lim8, x >= 0])
problem.solve(solver='glpk')  
print('Status: ', problem.status)
print('x: ', x.value)
print("Минимальные издержки на реализацию:")
print(problem.objective.value()[0])
print("Great")

Status:  optimal
x:  [ 0.00e+00]
[ 0.00e+00]
[ 1.20e+01]
[ 2.60e+01]
[ 4.70e+01]
[ 3.00e+01]
[ 3.70e+01]
[ 4.10e+01]
[ 0.00e+00]
[ 0.00e+00]
[ 0.00e+00]
[ 0.00e+00]
[ 0.00e+00]
[ 0.00e+00]
[ 0.00e+00]

Минимальные издержки на реализацию:
1250.0
Great


Таким образом, получили значение минимальных издержек на реализацию товара равное 1250 единиц. 

### *1.2. Выявим населенные пункты, требующие максимальный и минимальный рекламный бюджет:*

Данную задачу отнесем к задачам динамического программирования и в качестве управления выберем изменение распределения бюджета в зависимости от стоимости акции для каждого города.

### **Задание №2.**

В районе города, схема которого изображена на рис. 1, рассматривается возможность размещения пожарных участков (возможные точки размещения обозначены номерами, линии соответствуют дорогам, а закрашенные эллипсы - природным объектам).  
Стоимость размещения участка в каждой из точек указана в табл. 3.   
Требуется найти такое размещение участков, при котором стоимость была бы минимальна, но (манхеттенское) расстояние от каждого перекрестка до ближайшего участка было не более 3.


*рис. 1*

![Условие задачи](images/image1.jpg "почему..")

*табл. 3*

| Расположение | Стоимость, д.е. |
|--------------|-----------------|
| 1            | 15              |
| 2            | 15              |
| 3            | 12              |
| 4            | 20              |
| 5            | 14              |

### Формальная постановка:

$ F = 15x_1 + 15x_2 + 12x_3 + 20x_4 + 14x_5 \to min $

Для составления системы ограничений пронумеруем все перекрестки, всего их 12, учитывая наличие природных объектов. Отобразим полученную схему на рис. 2.

![Пронумерованные перекрестки](images/image2.jpg "почему..")

Теперь, исходя из данной схемы, составим ограничения по каждому перекрестку:   

$1: 0x_1 + 1x_2 + 1x_3 + 0x_4 + 0x_5 \geqslant 1$  
$2: 0x_1 + 1x_2 + 1x_3 + 1x_4 + 0x_5 \geqslant 1$  
$3: 0x_1 + 1x_2 + 1x_3 + 1x_4 + 0x_5 \geqslant 1$  
$4: 0x_1 + 1x_2 + 1x_3 + 1x_4 + 1x_5 \geqslant 1$  
$5: 1x_1 + 0x_2 + 1x_3 + 1x_4 + 0x_5 \geqslant 1$  
$6: 1x_1 + 0x_2 + 1x_3 + 1x_4 + 1x_5 \geqslant 1$  
$7: 1x_1 + 1x_2 + 1x_3 + 1x_4 + 1x_5 \geqslant 1$  
$8: 1x_1 + 0x_2 + 1x_3 + 1x_4 + 1x_5 \geqslant 1$  
$9: 1x_1 + 0x_2 + 1x_3 + 1x_4 + 1x_5 \geqslant 1$  
$10: 1x_1 + 0x_2 + 1x_3 + 1x_4 + 1x_5 \geqslant 1$  
$11: 1x_1 + 0x_2 + 1x_3 + 1x_4 + 1x_5 \geqslant 1$  
$12: 1x_1 + 0x_2 + 1x_3 + 1x_4 + 0x_5 \geqslant 1$  



In [116]:
x = variable(5, 'x')
F = (15 * x[0] + 15 * x[1] + 12 * x[2] + 20 * x[3] + 14 * x[4])

lim1 = (x[1] + x[2] >=1)
lim2 = (x[1] + x[2] + x[3] >=1)
lim3 = (x[1] + x[2] + x[3] >=1)
lim4 = (x[1] + x[2] + x[3] + x[4] >=1)
lim5 = (x[0] + x[2] + x[3] >=1)
lim6 = (x[0] + x[2] + x[3] + x[4] >=1)
lim7 = (x[0] + x[1] + x[2] + x[3] + x[4] >=1)
lim8 = (x[0] + x[2] + x[3] + x[4] >=1)
lim9 = (x[0] + x[2] + x[3] + x[4] >=1)
lim10 = (x[0] + x[2] + x[3] + x[4] >=1)
lim11 = (x[0] + x[2] + x[3] + x[4] >=1)
lim12 = (x[0] + x[2] + x[3] >=1)
problem = op(F, [lim1, lim2, lim3, lim4, lim5, lim6, lim7, lim8, lim9, lim10, lim11, lim12, x >= 0])
problem.solve(solver='glpk')  
print('Status: ', problem.status)
print("Результат:")
print('x: ', x.value)
print('F = ', problem.objective.value()[0])

# c = matrix([15, 15, 12, 20, 14], tc = 'd')

# G = matrix([[0, 0, 0, 0, -1, -1, -1, -1, -1, -1, -1, -1],
           #[-1, -1, -1, -1, 0, 0, -1, 0, 0, 0, 0, 0], 
           #[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1], 
           #[0, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1], 
           
           #[0, 0, 0, -1, 0, -1, -1, -1, -1, -1, -1, 0]], tc = 'd')
#h = matrix([-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1], tc='d')


Status:  optimal
Результат:
x:  [ 0.00e+00]
[ 0.00e+00]
[ 1.00e+00]
[ 0.00e+00]
[ 0.00e+00]

F =  12.0


Получили следующий результат: минимальная стоимость размещения участков равна 12 при размещении одного участка на точке 3. 